syntax-markdown-commonmark12/src/main/java/org/xwiki/contrib/rendering/markdown/commonmark12/internal/parser/DeepInlineHTMLPostProcessor.java

```diff
@@ -0,0 +1,110 @@
+import com.vladsch.flexmark.parser.block.NodePostProcessor;
+import com.vladsch.flexmark.parser.block.NodePostProcessorFactory;
+import com.vladsch.flexmark.util.ast.Document;
+import com.vladsch.flexmark.util.ast.Node;
+import com.vladsch.flexmark.util.ast.NodeTracker;
+import com.vladsch.flexmark.util.data.DataHolder;
+
+/**
+ * Detects and merges nodes that are content of inline HTML open and close tags.
+ * In a situation where a paragraph contains an open tag without any corresponding closing tag, the rest of the
+ * paragraph content will be treated as a single HtmlInline node.
+ *
+ * @version $Id$
+ * @since 8.9
+ */
+public class DeepInlineHTMLPostProcessor extends NodePostProcessor
+{
+    private static final String HTML_OPEN_PREFIX = "<";
+    private static final String HTML_CLOSE_PREFIX = "</";
+    private static final String HTML_SELF_SUFFIX = "/>";
+
+    /**
+     * Factory class for DeepInlineHTMLPostProcessor.
+     */
+    public static class Factory extends NodePostProcessorFactory
+    {
+        /**
+         * Factory constructor.
+         *
+         * @param options parser configuration
+         */
+        public Factory(DataHolder options)
+        {
+            super(false);
+            addNodes(HtmlInline.class);
+        }
+
+        @NotNull
+        @Override
+        public NodePostProcessor apply(@NotNull Document document)
+        {
+            return new DeepInlineHTMLPostProcessor();
+        }
+    }
+
+    @Override
+    public void process(@NotNull NodeTracker nodeTracker, @NotNull Node node)
+    {
+        // We store the number of nested open tags found during the processing and match it to the number of closing
+        // tags.
+        int toFind = 1;
+        int nFound = 0;
+
+        Node nextNode = node.getNext();
+        String inlineStartText = node.getChars().toString();
+
+        // We return early if we found a single closing tag, or a self-closing one.
+        if (inlineStartText.startsWith(HTML_CLOSE_PREFIX) || inlineStartText.endsWith(HTML_SELF_SUFFIX)) {
+            return;
+        }
+
+        // We keep the name of the tag until the first whitespace character.
+        String inlineStartTag = inlineStartText.split("\\s+")[0].substring(1);
+
+        while (nextNode != null && nFound < toFind) {
+            Node currentNode = nextNode;
+            nextNode = nextNode.getNext();
+
+            // We store the content of the processed nodes in the first HtmlInline node.
+            node.setChars(node.getChars().append(currentNode.getChars()));
+
+            if (currentNode instanceof HtmlInline) {
+                String currentNodeText = currentNode.getChars().toString();
+                if (currentNodeText.startsWith(HTML_OPEN_PREFIX + inlineStartTag)) {
+                    toFind++;
+                } else if (currentNodeText.startsWith(HTML_CLOSE_PREFIX + inlineStartTag)) {
+                    nFound++;
+                }
+            }
```

Comment:
What happens if there is something in between that is not HTML? Like if you have, e.g., `<strong> Text  _italic_ </strong>`? Also, wouldn't this logic be confused if you had something like `<a>link <abbr>HTML</a>`?

Reply:
For the first point, I assumed that it wouldn't make sense to include markdown in HTML, in markdown (I also considered the fact that it would not be supported in block macros either, so that was consistent). But looking at the commonmark specification, it is indeed valid but only for inline HTML. Now, to actually implement this, I can't really think of another way than converting the content of the inline macros to HTML silently, which is doable but I don't know if that's what we want.

For the second point, that's a good catch and an easy fix :+1:

Reply:
My idea for the first point was to just end the HTML macro before the inline syntax as a first step to not break the content. My understanding of the current code is that it wouldn't notice if there is Markdown syntax nested in the HTML and just ignore that Markdown syntax. First priority would be for me to preserve whatever there is in the syntax and not silently discard it.

Reply:
Since https://github.com/xwiki-contrib/syntax-markdown/pull/3/commits/79795cb03589e218ae88f251743990faad8dfca5, Markdown content between inline HTML tags is no longer ignored, but converted to HTML.
I agree this is not ideal since it can transform user input, but in the end I think it makes the most sense. The only alternative I see would be cutting the HTML macro in (at-least) three parts whenever it contains Markdown which would be very impractical in the WYSIWYG editor.